# Mosaico Musical

### Musical recommender by Alberto Antón as a final project for the Master in Data Science of KSchool


In [21]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import io
import os
import sys
import random

In [221]:
# display results to 3 decimal points, not in scientific notation, and thousands separator
pd.set_option("display.float_format", lambda x: "{:,.2f}".format(x))

In [22]:
# Set random seed
random.seed(666)

### Loading data

In [3]:
data_root = "data"

In [5]:
# Load training dataset
columns = ["user_id", "song_id", "num_plays"]
datafile = os.path.join(data_root, "train_triplets.txt")

data = pd.read_csv(datafile, 
                   sep=""\t", 
                   header = None,
                   names = columns)

In [6]:
# Let's get a glimpse of the data
data.head()

user_id             song_id  num_plays
0  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOAKIMP12A8C130995          1
1  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOAPDEY12A81C210A9          1
2  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBBMDR12A8C13253B          2
3  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBFNSP12AF72A0E22          1
4  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOBFOVM12A58A7D494          1

In [15]:
data.describe()

num_plays
count 48,373,586.00
mean           2.87
std            6.44
min            1.00
25%            1.00
50%            1.00
75%            3.00
max        9,667.00

In [16]:
# Let's analyze num_plays column a little deeper
data.num_plays.describe()

count   48,373,586.00
mean             2.87
std              6.44
min              1.00
25%              1.00
50%              1.00
75%              3.00
max          9,667.00
Name: num_plays, dtype: float64

In [17]:
# most of the songs have been played only one time, and there are very large outliers, 
# so we will not be using num_plays field.

We already have the listenings dataset, now let's load the song information dataset

In [60]:
columns = ["foo", "song_id", "artist", "title"]
datafile = os.path.join(data_root, "unique_tracks.txt")

all_songs = pd.read_csv(datafile, 
                        header = None,
                        sep = "<SEP>",
                        names = columns,
                        usecols = ["song_id", "artist", "title"],
                        encoding =  "utf-8",
                        engine = "python")

In [61]:
all_songs.sample(10)

song_id                                 artist  \
299918  SOVSONN12A8C133578               The Charlie Daniels Band   
600203  SOCGOAR12A6D4F5EA3                    The Beautiful South   
690933  SOKLENT12A8C1401D7            James Galway;Daniel Measham   
413968  SOMAEWN12AB017E37F                              zebrahead   
425908  SOKHRQI12A8C13F53E  Cam'Ron / Juelz Santana / Jimmy Jones   
934697  SOXNBUL12AC468D570                                Slowpho   
117534  SOZFNVI12AB018887D                                 Amanda   
463207  SOIRJRE12A8C135117             Southern Tenant Folk Union   
787914  SOBATOZ12AB017FC6F                           An Albatross   
681144  SOQAMQQ12A8AE4866D                           Annie Lennox   

                                title  
299918                      Tennessee  
600203   Good As Gold (Stupid As Mud)  
690933               Waltzing Matilda  
413968  All for none and none for all  
425908              Come Home With Me  
934697     Dial (Stupid Høghus Remix)  
117534                     Ingenstans  
463207       Southern Folk Theme In A  
787914         Dimensional Gymnastics  
681144    The Saddest Song (I've Got)

In [62]:
all_songs.describe()

song_id           artist   title
count              1000000          1000000  999985
unique              999056            72665  702000
top     SONBEKD12AB01894DC  Michael Jackson   Intro
freq                     3              194    1511

In [63]:
# There seems to be repeated some songs. Lest's find out
all_songs[all_songs.duplicated(subset=['song_id'], keep=False)].sort_values("song_id").head(10)

song_id                                       artist  \
963681  SOAAEFC12AB01852F1  Tineke Schouten/Linda De Mol/Franklin Brown   
304966  SOAAEFC12AB01852F1                              Tineke Schouten   
572832  SOACGAQ12A58A79805                               Arctic Monkeys   
141269  SOACGAQ12A58A79805                               Arctic Monkeys   
15028   SOADFGH12A6D4F74F7                        Red Hot Chili Peppers   
522461  SOADFGH12A6D4F74F7                        Red Hot Chili Peppers   
68224   SOADYVX12A8A9D9462                                      Rihanna   
655219  SOADYVX12A8A9D9462                                      Rihanna   
182693  SOAEIFW12A8C1391E4                              Franz Ferdinand   
525920  SOAEIFW12A8C1391E4                              Franz Ferdinand   

                                                 title  
963681                                   De Tongbreker  
304966  De Tongbreker (Tineke Schouten & Linda de Mol)  
572832                               Fire And The Thud  
141269                               Fire And The Thud  
15028               Falling Into Grace (Album Version)  
522461              Falling Into Grace (Album Version)  
68224                                          We Ride  
655219                                         We Ride  
182693                                         Michael  
525920                                         Michael

In [64]:
# Let's remove those duplicates
all_songs.drop_duplicates(subset = "song_id", inplace = True)
all_songs.describe()

song_id       artist   title
count               999056       999056  999041
unique              999056        72652  701922
top     SOEATPZ12AB0182EA3  Johnny Cash   Intro
freq                     1          191    1511

Now count and unique show the same ammount

In [29]:
# There is information of about one million songs. Let's see hoy many of these songs
# are in the training dataset
data.song_id.unique().shape[0]

384546

We don't need information for so many songs, so let's create a dataframe with information only on the songs that are in the training dataframe.

In [65]:
# Let's keep only the songs that are in data
unique_songs_df = pd.DataFrame(data.song_id.unique(), columns = ["song_id"])

In [66]:
songs = unique_songs_df.merge(all_songs, on = "song_id", how = "left")[["song_id", "artist", "title"]]

In [67]:
songs.describe()

song_id        artist   title
count               384546        384546  384542
unique              384546         42055  306720
top     SOLSHRD12AB018BF73  Beastie Boys   Intro
freq                     1           136     526

We can now release some memory discarding the all_songs dataframe

In [68]:
all_songs = None

If we want to recommend something to a new user, either we know something about him or we can't but recommend the most popular songs.

Let's go for the first option, so we'll choose some songs and make the recommender to select some new ones for us, but first, we are going to create some helper functions.

In [69]:
# Function that shows the songs of an artist
def songs_by(artist_name):
    return songs[songs["artist"] == artist_name][["song_id", "title"]]

In [70]:
# Let's try to find some rock songs an see the recommendation...
songs.artist.sample(25)

289281                                 Adultnapper
84521                                   Dusminguet
364797                            Thought Industry
243807                                  Juaninacka
38834                                         Koop
242478                               The Oppressed
202995                                        Malu
53805                                 Stevie Nicks
41188                             The Aqua Velvets
38899                                Easton Corbin
312948                 Dolly Parton & Kenny Rogers
213235                                Lizzy Borden
178214                                   All-4-One
288538                                 Alex Baroni
54787                                          Ayo
245196                   The Strange Fruit Project
217254                                   The Itals
244223    Steve Wynn & The Dragon Bridge Orchestra
21364                                         Salt
319667                         

In [72]:
songs_by('Metallica')

song_id                                   title
1146    SOOEEPE12A8AE459A4                      The Unforgiven III
1407    SOZATKE12A6D4F5915                                   2 X 4
1594    SOGAUIQ12A6D4F8262                          Hit The Lights
1641    SOUGBIM12A6D4F8247                       The Four Horsemen
1713    SOCHYVZ12A6D4F5908                           Enter Sandman
1787    SOZDGEW12A8C13E748                                     One
1794    SOGMBXD12A6D4F5920                                  Ronnie
4702    SOJSRYJ12A6D4F824C                            Phantom Lord
4712    SOMTBXX12AF729F5A6                              Am I Evil?
4726    SORIEXB12A6D4F824D                              No Remorse
4727    SORMSJW12A6D4F824F                           Metal Militia
4739    SOXKNEB12A6D4F824A             (Anaesthesia) Pulling Teath
4788    SOKIORP12A6D4F5917                         Until It Sleeps
6285    SOSHZYN12A6D4F5925                       The Unforgiven II
6340    SOSJRJP12A6D4F826F                       Master Of Puppets
9179    SOITRTA12A6D4F8261                      Ride The Lightning
9214    SOUZJMM12A6D4F5911                     The God That Failed
10211   SOMCHKO12A58A7849A                          Shoot Me Again
10468   SOCIRNW12A6D4F5948                      Stone Dead Forever
11517   SOMAIEW12A6D4F591E                               Mama Said
11551   SOSWDMO12A8AE45996                The Day That Never Comes
11727   SOPNXGM12A6D4F591B                                    Cure
11733   SOUOIWT12A8C142D8C                             Bleeding Me
14501   SOAEEMQ12A8C142D36                     The Ecstasy Of Gold
15808   SOKYMHM12A6D4F8270            The Thing That Should Not Be
16416   SOLNDXN1280EC900A1                                Loverman
16418   SONCQKA12A6D4F825E                             Damage Inc.
16432   SOUYXJU12A8AE45993                  Broken_ Beat & Scarred
19392   SOAWFKO129F06933A7               Welcome Home (Sanitarium)
19418   SOQDMXT12A6D4F8255                           Fade To Black
...                    ...                                     ...
27164   SOCWSNN12A6D4F825C                    Some Kind Of Monster
30843   SOIHJML12A6D4F5944                                 So What
31928   SOVXERM12A8C142468                             Outlaw Torn
31935   SOVZVGK12A8C142DB7                     Wherever I May Roam
31950   SOFKBKR12A8AE4598E                     The End Of The Line
32308   SOGHOFP12A8C142D78                          No Leaf Clover
32554   SOOXPVT12A58A78491                                My World
34341   SODAWDD12A6D4F825A                 For Whom The Bell Tolls
34344   SODEMBR12A6D4F5929                                Bad Seed
34473   SOKFLTD12A8C142D54                         Of Wolf And Man
45013   SOGQPOC12A6D4F5938                          Tuesday's Gone
50925   SOWJMGE12A6D4F58B7                                 Battery
50991   SOYCNOY12A6D4F5935                           Mercyful Fate
51925   SOBWDJL12A6D4F825B                          Creeping Death
51963   SOSOWFW12A58A7B97C                            Killing Time
57140   SOWVCXK12A6D4F591F                            Thorn Within
64720   SOZRSGF12A8C13E458                           Enter Sandman
68545   SOKMBHN12A6D4F5931                          Sabbra Cadabra
75819   SOJPYOR12A6D4F5928                         Carpe Diem Baby
83622   SONKMPN12A6D4F5927                                 Slither
83624   SOVRLVW12A81C1FFE9                            Dirty Window
95691   SOBRWBK12A8C13E784                  ...And Justice For All
99508   SOWNJYY12A81C20010  The Unnamed Feeling (Explicit Version)
134415  SOPXTQW12A6D4F58B8        The Thing That You Should Not Be
140110  SOUXLOB12A8C142D42                   The Call Of The Ktulu
168090  SOPFRZX12A6D4F8263                    Some Kind Of Monster
227325  SOEWAYK12A8C142D83                           Devil's Dance
265119  SOBCPOL12AB01809A5                          Creeping Death

We list some groups. When we find one we like, list its songs and get the song id. We do this until we have a small dataset of songs we like.

This is our metalhead selection: 

| Artist | song_id   |
|------|------|
| Thrice | SORJRTI12A6D4F7D67 |
| Clawfinger | SOSOPGB12A8C13C185 |
| Rammstein | SOSYHME12A8C135DD8 | 
| Rancid | SOSBJSU12A8C138469 | 
| Against Me! | SONJQZM12A6D4FBE30 |
| Millencolin | SOZVBUH12A8AE4745C |
| Van Halen | SOVMGEX12AC9070FF2 |
| Led Zeppelin | SOEHJKJ12A8C13CA4D |
| Staind | SOUNBBX12A6D4F338E |
| Monster Magnet | SOJKARY12A6701ED3F |
| Whitesnake | SOPCNEA12A67ADF48B |
| Puddle Of Mudd | SOTQVSE12A6D4F8200 |
| Green Day | SOTNYYH12A6701F94B |
| Metallica | SOSJRJP12A6D4F826F |

In [74]:
selected_songs = ["SORJRTI12A6D4F7D67", "SOSOPGB12A8C13C185", "SOSYHME12A8C135DD8", "SOSBJSU12A8C138469", \
                 "SONJQZM12A6D4FBE30", "SOZVBUH12A8AE4745C", "SOVMGEX12AC9070FF2", "SOEHJKJ12A8C13CA4D", \
                 "SOUNBBX12A6D4F338E", "SOJKARY12A6701ED3F", "SOPCNEA12A67ADF48B", "SOTQVSE12A6D4F8200", \
                 "SOTNYYH12A6701F94B", "SOSJRJP12A6D4F826F"]

In [86]:
# Now let's find other users that listened to these songs. we need it as a dataframe, so we'll use reset_index()
similar_users = data[data["song_id"].isin(selected_songs)]['user_id'].reset_index()

In [89]:
similar_users.head()

index                                   user_id
0   6041  bd64f193f0f53f09d44ff48fd52830ff2fded392
1   7067  a520488fcf049bbb5cd847cfa4f884c740692780
2   7970  0ef42a19efb74d0a05c308d00636c8d8d41bec0c
3   8466  7661038e3e655fd31961ad18aea13dded963eedf
4   9524  12497e138741a0b94bb36a14bef32c9d0ee20fec

In [90]:
similar_users.user_id.describe()

count                                        36391
unique                                       35253
top       52542a715ba72e52eec99b277a42532c88615469
freq                                             4
Name: user_id, dtype: object

We can see that most of the users only have played one of the soungs of our selection, and that the one who has listened to more of our songs has played only 4.

In [80]:
# Let's see how many of our songs those users have played

data[data["song_id"].isin(selected_songs)].groupby(["user_id"]).size().sort_values(ascending = False)

user_id
8fc187765f25645e802bd5137f641c8de7df17b8    4
52542a715ba72e52eec99b277a42532c88615469    4
be59c5b281f8b714c4d4d4bfb877715a93b3c64d    4
58c846a9d19a9345bffe62b212436cb49363278a    3
a883218d1e6171d4913b1dec6c083eb3fea5f914    3
b73bc9b4732c8edf790e257df7395973f8d085ef    3
3e6dd161e97e7bd0e20986e7f5e391e5d24e0a62    3
b67f2d3bea6a313bc55695517cc9b38ff5f920fa    3
07e8066fc9c82f5e700023f3c963117e874e0188    3
3f52fdf255f7043eb170a49606bebe14f6c7a08a    3
ed824982bb5d17465708f5bfbd8589af81ad4de0    3
217b76adb93cdb5d221408ad9f9c5c244a65b038    3
f106f63e74ba0648ed27e2fd59094a53c8c9c534    3
78fb080641b1b1f9b85ceffd9c1686eb8db7c765    3
748096044d04f6736c6921203f711f57fe6e31ee    3
00fc9d7d12f74bcd93fa787cc26a9c61a0904ac7    3
b161e27efcd0135dabd0cc2cfea477498667b191    3
53175b45ba820a33ac8f833a85a986a7c0f7d3d4    3
b60c2902ab24963f33d8a431bee8676a14ceb003    3
8c24607fcd3b2ca28a8eb5924c7c26d8d40e82c4    3
127c8ac775ebdce42de94ff5783ab8d8e333711f    3
58dc40ef3b13f15b889f72d6e3

Now that we know the users that have played our music, we need to know what other songs they have listened.

In [91]:
prediction_df = similar_users.merge(data[~data.song_id.isin(selected_songs)], on = 'user_id')

In [92]:
prediction_df.head()

index                                   user_id             song_id  \
0   6041  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOACIPG12A8AE47E1C   
1   6041  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOAHEEC12A6BD4DAA4   
2   6041  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOAKQBB12A8C1413A0   
3   6041  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOAVFMF12A6D4F92E6   
4   6041  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOBOLEI12A58A7E386   

   num_plays  
0          1  
1          1  
2          1  
3          1  
4          1

In [94]:
prediction_df.describe()

index    num_plays
count  3,233,753.00 3,233,753.00
mean  24,217,294.64         2.99
std   14,002,352.72         6.43
min        6,041.00         1.00
25%   11,973,702.00         1.00
50%   24,413,248.00         1.00
75%   36,386,260.00         3.00
max   48,373,378.00     2,213.00

In [95]:
prediction_df.user_id.describe()

count                                      3233753
unique                                       35253
top       4e73d9e058d2b1f2dba9c1fe4a8f416f9f58364f
freq                                          4623
Name: user_id, dtype: object

Now we have all the songs our similar users have played. We only have to select the most popular ones. We can to this in two ways:

1. Order the songs by play count (repetition).
2. Order the songs by the number of users that have played them (popularity).

The repetition is very sensitive to outliers. As we saw at the beginning, the maximum value of play_count is 9,667. Everytime this user is selected, that song would be the first in our recommendation, so we are going to go with the popularity option.

In [96]:
# Let's now find the most played songs in the prediction dataframe. This would be our recommendation
predicted_songs = prediction_df.\
                    groupby(['song_id']).\
                    size().\
                    sort_values(ascending = False).\
                    head(20).\
                    to_frame("popularity").\
                    reset_index()

In [97]:
predicted_songs.head()

song_id  popularity
0  SOEGIYH12A6D4FC0E3        8851
1  SOAUWYT12A81C206F1        8014
2  SOSXLTC12AF72A7F54        7285
3  SOBONKR12A58A7A7E0        6980
4  SOFRQTD12A81C233C0        6086

In [98]:
# Let's see the songs in a human readable way
predicted_songs.merge(songs, on = 'song_id')[['artist', 'title']]

artist  \
0   Barry Tuckwell/Academy of St Martin-in-the-Fie...   
1                                               Björk   
2                                       Kings Of Leon   
3                                       Dwight Yoakam   
4                                            Harmonia   
5                                     Alliance Ethnik   
6                                             Cartola   
7                                         OneRepublic   
8                                  Lil Wayne / Eminem   
9                              Florence + The Machine   
10                                    Alliance Ethnik   
11                      J. Karjalainen & Mustat Lasit   
12                                   Five Iron Frenzy   
13                                         Todd Barry   
14                                      Lonnie Gordon   
15                                         The Crests   
16                                           Tub Ring   
17                                        Future Rock   
18                                Angels and Airwaves   
19                                 Charttraxx Karaoke   

                                                title  
0   Horn Concerto No. 4 in E flat K495: II. Romanc...  
1                                                Undo  
2                                             Revelry  
3                                      You're The One  
4                                       Sehr kosmisch  
5                                          Représente  
6                                            Tive Sim  
7                                             Secrets  
8                                      Drop The World  
9                      Dog Days Are Over (Radio Edit)  
10                              Sincerité Et Jalousie  
11                              Sinisten tähtien alla  
12                                             Canada  
13                             Sugar Ray (LP Version)  
14  Catch You Baby (Steve Pitron & Max Sanna Radio...  
15                                         16 Candles  
16                                            Invalid  
17                                              Gears  
18                                           The Gift  
19                                          Fireflies

Mmmmm... There doesn't seem to be much metal in that list, does it? The result is totally different from the set of songs we chose. Let's investigate. First we are going to take a look at the most popular songs of the entire dataset to see it it's similar to this.

In [99]:
# Let's find out the most played songs in the whole dataset
most_popular_songs = data.\
                    groupby(['song_id']).\
                    size().\
                    sort_values(ascending = False).\
                    head(20).\
                    to_frame("popularity").\
                    reset_index()

In [101]:
most_popular_songs.merge(songs, on = 'song_id')[['song_id', 'artist', 'title', 'popularity']]

song_id                                             artist  \
0   SOFRQTD12A81C233C0                                           Harmonia   
1   SOAUWYT12A81C206F1                                              Björk   
2   SOAXGDH12A8C13F8A1                             Florence + The Machine   
3   SOBONKR12A58A7A7E0                                      Dwight Yoakam   
4   SOSXLTC12AF72A7F54                                      Kings Of Leon   
5   SONYKOW12AB01849C9                                        OneRepublic   
6   SOEGIYH12A6D4FC0E3  Barry Tuckwell/Academy of St Martin-in-the-Fie...   
7   SOLFXKT12AB017E3E0                                 Charttraxx Karaoke   
8   SODJWHY12A8C142CCE                                              Train   
9   SOFLJQZ12A6D4FADA6                                            Cartola   
10  SOUSMXX12AB0185C24                          Usher featuring will.i.am   
11  SOWCKVR12A8C142411                                      Kings Of Leon   
12  SOUVTSM12AC468F6A7                                 Lil Wayne / Eminem   
13  SOTWNDJ12A8C143984                                              Train   
14  SOPUCYA12A8C13A694                                   Five Iron Frenzy   
15  SOHTKMO12AB01843B0                                      Lonnie Gordon   
16  SOKLRPJ12A8C13C3FE                                           Coldplay   
17  SOCVTLJ12A6310F0FD                                           Coldplay   
18  SOPPROJ12AB0184E18                          Kid Cudi / MGMT / Ratatat   
19  SOBOUPA12A6D4F81F1                                    Alliance Ethnik   

                                                title  popularity  
0                                       Sehr kosmisch      110479  
1                                                Undo       90476  
2                      Dog Days Are Over (Radio Edit)       90444  
3                                      You're The One       84000  
4                                             Revelry       80656  
5                                             Secrets       78353  
6   Horn Concerto No. 4 in E flat K495: II. Romanc...       69487  
7                                           Fireflies       64229  
8                                    Hey_ Soul Sister       63809  
9                                            Tive Sim       58610  
10                                                OMG       53260  
11                                       Use Somebody       52080  
12                                     Drop The World       51022  
13                                           Marry Me       47011  
14                                             Canada       46078  
15  Catch You Baby (Steve Pitron & Max Sanna Radio...       46077  
16                                      The Scientist       45495  
17                                             Clocks       42535  
18                   Pursuit Of Happiness (nightmare)       41811  
19                              Sincerité Et Jalousie       41093

Ok. Now we see a couple of things about our recommender:

1. It is heavily influenced by popularity. There are 35,253 users that have listened to at least one of our songs, and the vast majority has listened to just one of them, meaning that they are not very similar to us, so their recomendation is almost random. This forces us to give more importance to the songs of the users more similar to us.

2. Harmonia the most popular artist? Bjök the second? Dwight Yoakam the fourth? This is impossible in real life. This dataset must be rotten, but anyway, to our recommendator they are just names.

Now we are going to implement the scoring funtion to put more weight on the songs of the most similar to us users than on the least similar. We will create a score for each song based on how similar to us is the user. Later we will sum all the scores and return an ordered by score list that will be our new recommendation.

First we have to know every user similarity to us:

In [103]:
# Create a Dataframe with the number of similar songs by user
user_similarity = data[data["song_id"].isin(selected_songs)].groupby(['user_id']).size().to_frame("similarity")

In [105]:
# Let's see the ten most similar to us users
user_similarity.sort_values('similarity', ascending = False).head(10)

similarity
user_id                                             
8fc187765f25645e802bd5137f641c8de7df17b8           4
52542a715ba72e52eec99b277a42532c88615469           4
be59c5b281f8b714c4d4d4bfb877715a93b3c64d           4
217b76adb93cdb5d221408ad9f9c5c244a65b038           3
33825a5d5b1b2ea935a9fc2f4a3cbf8e97e6280a           3
07e8066fc9c82f5e700023f3c963117e874e0188           3
b31888d485ddff26572ffdab1c947bcc067ff3a1           3
ed824982bb5d17465708f5bfbd8589af81ad4de0           3
53175b45ba820a33ac8f833a85a986a7c0f7d3d4           3
3e6dd161e97e7bd0e20986e7f5e391e5d24e0a62           3

In [111]:
# Now we are going to create a new dataframe with the user_id, song_id and similarity excluding the songs
# we have selected for our recommendation.
prediction_df = similar_users.\
    merge(data[~data.song_id.isin(selected_songs)], on = 'user_id').\
    merge(user_similarity, on = "user_id")[["user_id", "song_id", "similarity"]]

In [112]:
prediction_df.sample(10)

user_id             song_id  \
3047192  5ad93cdc05cf1431c04d7e809df3934d980b5a5f  SOQGVCS12AF72A078D   
2114466  8a5fadcd48a1cbffc0a67795acbbb624b45f6ad8  SOWWWSD12A58A801F9   
3039753  15206b4183c1b9effadf75247bbf51cfbb387a72  SOXXICK12A6D4F67B9   
1838791  3487d95158d8ec4bc6b10ffb2db39593eccc556a  SOOONEP12A8AE483EC   
2351891  550eda18938d8cf21990ca15685c5a050dbc7c75  SOOPQAY12A6D4F5926   
2818013  8a9dc382f60363f76354c857e0a97a98b9bc74e7  SOQOLYG12AF729EAAC   
1468193  872821d140b7f318205357a8823ee6177aea4c23  SOACIPG12A8AE47E1C   
173699   bd1edf96f947af790365f7378d2e894ff4f7b72e  SOTJCZJ12A58A7E54B   
576247   2904f00329117109159e6831b16b8e9dc866480a  SOUSMXX12AB0185C24   
2432536  592aeeaa43b2af7a5670085eefc0fb9f4987f23f  SOLVRYB12A6D4FBA1D   

         similarity  
3047192           1  
2114466           2  
3039753           1  
1838791           1  
2351891           1  
2818013           3  
1468193           1  
173699            1  
576247            2  
2432536           1

In [115]:
# now we have to create a DF with the score of every song that we'll user later to order the most relevant songs
# for a specific user
max_similarity = pred_df.similarity.max()

In [116]:
total_plays_similar_users = pred_df.shape[0]

In [250]:
# Anton's scoring funtion that ranks songs by similarity to the user
def song_scoring(similarity):
    return (similarity / max_similarity) ** (max_similarity - (similarity - 1))

In [251]:
# Constant values for normalization
max_sim = song_scoring(max_similarity)
min_sim = song_scoring(1)

In [257]:
# Let's apply Anton's scoring funtion to the prediction dataframe
prediction_df['score'] = prediction_df['similarity'].map(lambda x: song_scoring(x))

In [258]:
prediction_df.head(10)

user_id             song_id  similarity  \
0  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOACIPG12A8AE47E1C           1   
1  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOAHEEC12A6BD4DAA4           1   
2  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOAKQBB12A8C1413A0           1   
3  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOAVFMF12A6D4F92E6           1   
4  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOBOLEI12A58A7E386           1   
5  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOBYNII12A58291CDC           1   
6  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOCALRI12A58A7BBC5           1   
7  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOCHYVZ12A6D4F5908           1   
8  bd64f193f0f53f09d44ff48fd52830ff2fded392  SODQGBE12A6D4F6BAB           1   
9  bd64f193f0f53f09d44ff48fd52830ff2fded392  SOEAQHH12A58A78F59           1   

   score  
0   0.00  
1   0.00  
2   0.00  
3   0.00  
4   0.00  
5   0.00  
6   0.00  
7   0.00  
8   0.00  
9   0.00

In [259]:
# Let's summ the score of every song and order it
predicted_songs = prediction_df.\
                    groupby(['song_id'])['score'].\
                    sum().\
                    sort_values(ascending = False).\
                    head(20).\
                    to_frame("popularity")

In [260]:
predicted_songs.head(20)

popularity
song_id                       
SOTNHIP12AB0183131      102.01
SOEGIYH12A6D4FC0E3       88.20
SOZDGEW12A8C13E748       84.81
SOLRGNF12AB0187CF4       83.88
SOFGIVB12A6D4F5923       78.56
SOITRTA12A6D4F8261       76.16
SOPQLBY12A6310E992       74.57
SOSXLTC12AF72A7F54       74.00
SOCHYVZ12A6D4F5908       73.45
SOAUWYT12A81C206F1       72.64
SOLLNTU12A6701CFDC       69.46
SOAUFOF12AB0180C65       69.05
SOQGVCS12AF72A078D       68.48
SOGHFDV12A6D4F7E0D       68.27
SOSFPMR12A8C13D44A       68.14
SOFRQTD12A81C233C0       66.38
SOBONKR12A58A7A7E0       65.70
SOTOXBH12AAF3B5B96       65.41
SOQLUTQ12A8AE48037       63.99
SOCKSGZ12A58A7CA4B       63.22

In [261]:
# Let's see the songs in a human readable way
predicted_songs.merge(songs, on = 'song_id')[['song_id', 'artist', 'title']].head(20)

song_id                                             artist  \
0   SOTNHIP12AB0183131                     Kid Cudi / Kanye West / Common   
1   SOEGIYH12A6D4FC0E3  Barry Tuckwell/Academy of St Martin-in-the-Fie...   
2   SOZDGEW12A8C13E748                                          Metallica   
3   SOLRGNF12AB0187CF4                                       Simon Harris   
4   SOFGIVB12A6D4F5923                     Metallica / Marianne Faithfull   
5   SOITRTA12A6D4F8261                                          Metallica   
6   SOPQLBY12A6310E992                                          Radiohead   
7   SOSXLTC12AF72A7F54                                      Kings Of Leon   
8   SOCHYVZ12A6D4F5908                                          Metallica   
9   SOAUWYT12A81C206F1                                              Björk   
10  SOLLNTU12A6701CFDC                                       3 Doors Down   
11  SOAUFOF12AB0180C65                                  The White Stripes   
12  SOQGVCS12AF72A078D                                      Guns N' Roses   
13  SOGHFDV12A6D4F7E0D                                          Rammstein   
14  SOSFPMR12A8C13D44A                                          Rammstein   
15  SOFRQTD12A81C233C0                                           Harmonia   
16  SOBONKR12A58A7A7E0                                      Dwight Yoakam   
17  SOTOXBH12AAF3B5B96                                          Rammstein   
18  SOQLUTQ12A8AE48037                                       Foo Fighters   
19  SOCKSGZ12A58A7CA4B                                        Linkin Park   

                                                title  
0                                        Make Her Say  
1   Horn Concerto No. 4 in E flat K495: II. Romanc...  
2                                                 One  
3                                      Sample Track 2  
4                                  The Memory Remains  
5                                  Ride The Lightning  
6                                    Creep (Explicit)  
7                                             Revelry  
8                                       Enter Sandman  
9                                                Undo  
10                                         Kryptonite  
11                                  Seven Nation Army  
12                                      Paradise City  
13                                            AMERIKA  
14                                         Feuer Frei  
15                                      Sehr kosmisch  
16                                     You're The One  
17                                           Ich Will  
18                                      The Pretender  
19               Bleed It Out [Live At Milton Keynes]

In [ ]:
# Sacar las canciones más oídas de los que tienen un 4 de similitud a ver si se parecen a esto...